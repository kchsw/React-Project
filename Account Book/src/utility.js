export const LIST_VIEW = 'list'
export const CHART_VIEW = 'chart'
export const TYPE_INCOME = 'income'
export const TYPE_OUTCOME = 'outcome'
export const padLeft = (n) => {
    return n < 10 ? '0' + n : n
}

export const range = (size, startAt = 0) => {
    const arr = []
    for(let i = 0; i < size; i++){
        arr[i] = startAt + i
    }
    return arr
}

export const parseDate = (str) => {
    const date = str? new Date(str) : new Date()
    return {
        year: date.getFullYear(),
        month: date.getMonth() + 1
    }

}

export const Colors = {
    blue: '#347eff',
    deepBlue: '#61dafb',
    green: '#28a745',
    red: '#dc3545',
    gray: '#555',
    lightGray: '#efefef',
    white: '#fff'
}