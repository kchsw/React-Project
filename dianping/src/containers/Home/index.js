import React, { Component } from 'react';
import Category from './components/Category'
import Headline from './components/Headline'
import Discount from './components/Discount'
import LikeList from './components/Likelist'
import HomeHeader from './components/HomeHeader'
import Banner from './components/Banner'
import Activity from './components/Activity'
import Footer from '../../components/Footer/index.js'

import { bindActionCreators } from "redux";
import { connect } from "react-redux";
import { actions as homeActions, getLikes, getDiscounts, getPageCountOfLikes } from '../../redux/modules/home'
class Home extends Component {
    state = {  }
    render() { 
        const { discounts, likes, pageCount } = this.props
        return ( 
            <>  
                <HomeHeader />
                <Banner dark />
                <Category />
                <Headline />
                <Activity />
                <Discount data={discounts}/>
                <LikeList 
                    data={likes} 
                    pageCount={pageCount}
                    fetchData = {this.fetchMoreLikes}
                />
                <Footer />
           </>
        );
    }
    componentDidMount() {
        this.props.homeActions.loadDiscounts()
    }
    fetchMoreLikes = () => {
        this.props.homeActions.loadLikes()
    }
}

const mapStateToProps = (state, props) => {
    return {
        likes: getLikes(state),
        discounts: getDiscounts(state),
        pageCount: getPageCountOfLikes(state)
    };
  };
  
const mapDispatchToProps = dispatch => {
    return {
        homeActions: bindActionCreators(homeActions, dispatch)
    };
};
 
export default connect(mapStateToProps, mapDispatchToProps)(Home);