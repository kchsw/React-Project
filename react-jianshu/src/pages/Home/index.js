import React, { Component } from 'react';
import Topic from './components/Topic'
import List from './components/List'
import Recommend from './components/Recommend'
import Writer from './components/Writer'
import { connect } from 'react-redux'
import { InitHomeList, toggleTopShow } from './store/actionCreators'
import {
    HomeWrapper,
    HomeLeft,
    HomeRight,
    BackTop
} from './style'
class Home extends Component {
    constructor(props) {
        super(props);
        this.state = {  }
    }
    handleScrollTop(){
        window.scroll(0, 0)
    }
    render() { 
        return (  
            <HomeWrapper>
                <HomeLeft>
                    <a href="/">
                    <img className="banner-img" src="https://upload.jianshu.io/admin_banners/web_images/4686/b205842c3dd6736c233b052411fe1211f164fcaf.png?imageMogr2/auto-orient/strip|imageView2/1/w/1250/h/540"/>
                    </a>
                    <Topic></Topic>
                    <List></List>
                </HomeLeft>
                <HomeRight>
                    <Recommend></Recommend>
                    <Writer></Writer>
                </HomeRight>
                {
                    this.props.showScroll ? <BackTop onClick={this.handleScrollTop}>回到顶部</BackTop> : null
                }
                
            </HomeWrapper>
        );
    }
    componentDidMount(){
        this.props.initHomeData()
        this.bindEvents()
    }

    bindEvents(){
        window.addEventListener('scroll', this.props.toggleScrollShow)
    }

    componentWillUnmount(){
        window.removeEventListener('scroll', this.props.toggleScrollShow)
    }
}

const mapState = (state) =>{
    return {
        showScroll: state.getIn(['home', 'showScroll'])
    }
}

const mapDispatch = (dispatch) => {
    return {
        initHomeData(){
            dispatch(InitHomeList())
        },
        toggleScrollShow(e){
            document.documentElement.scrollTop > 200 ? dispatch(toggleTopShow(true)) : dispatch(toggleTopShow(false))
        }
    }
}
export default connect(mapState, mapDispatch)(Home);