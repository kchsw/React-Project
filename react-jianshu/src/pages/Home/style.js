import styled from 'styled-components'

export const HomeWrapper = styled.div`
    width: 960px;
    overflow: hidden;
    margin: 0 auto;
`

export const HomeLeft = styled.div`
    margin-left: 15px;
    width: 625px;
    padding-top: 30px;
    float: left;
    .banner-img{
        width: 625px;
        height: 270px;
    }
`

export const HomeRight = styled.div`
    width: 240px;
    float: right;
`
export const TopicWrapper = styled.div`
    padding: 20px 0 10px 0;
    overflow: hidden;
    margin-right: -18px;
`
export const TopicItem = styled.div`
    float: left;
    height: 32px;
    line-height: 32px;
    background: #f7f7f7;
    font-size: 14px;
    color: #000;
    border: 1px solid #dcdcdc;
    border-radius: 4px;
    padding-right: 10px;
    margin: 0 18px 18px 0;
    .topic-pic{
        height: 32px;
        width: 32px;
        display: block;
        float: left;
        margin-right: 10px;
    }
`
export const TopicMore = styled.a.attrs({
    href: '/'
})` 
    text-decoration: none;
    display: block;
    float: left;
    height: 32px;
    line-height: 32px;
    font-size: 13px;
    color: #000;
`