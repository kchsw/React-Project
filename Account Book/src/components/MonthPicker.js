import React, { Component } from 'react';
import PropTypes from 'prop-types'
import { padLeft, range } from '../utility'

class MonthPicker extends Component {
    constructor(props) {
        super(props)
        this.state = {
            isOpen: false,
            selectedYear: this.props.year
        }
    }
    componentDidMount() {
        document.addEventListener('click', this.handleClick, false)
    }
    componentWillUnmount() {
        document.removeEventListener('click', this.handleClick, false)
    }
    handleClick = (event) => {
        if (this.node.contains(event.target)) {
          return;
        }
        this.setState({
          isOpen: false,
        })
    }
    toggleDropdown = (e) =>{
        e.preventDefault();
        this.setState({
            isOpen: !this.state.isOpen
        })
    }
    selectYear = (e, year) => {
        e.preventDefault();
        this.setState({
            selectedYear: year
        })
    }
    selectMonth = (e, month) => {
        e.preventDefault();
        this.setState({
            isOpen: false
        })
        this.props.onChange(this.state.selectedYear, month)
    }
    render() { 
        const { year, month } = this.props
        const { isOpen, selectedYear } = this.state
        const monthRange = range(12, 1)
        const yearRange = range(9, -4).map(i => i + year)
        return (
            <div className="dropdown position-relative" ref={(ref) => {this.node = ref}}>
                <h6>选择月份</h6>
                <button className="btn btn-sm btn-secondary dropdown-toggle"
                    onClick={this.toggleDropdown}
                >
                    {`${year}年${padLeft(month)}月`}
                </button>
                {
                    isOpen && 
                    <div className="row border position-absolute" style={{background: '#fff', zIndex: 10, width: '100%'}}>
                        <div className="col border-right years-range">
                           {
                               yearRange.map((yearNumber, index) =>
                                <a key={index} 
                                    className={selectedYear === yearNumber ? "dropdown-item active" : "dropdown-item"}
                                    href="#"
                                    onClick={(e) => {this.selectYear(e, yearNumber)}}
                                >
                                   {yearNumber}年
                                </a>
                                )
                           }
                        </div>
                        <div className="col months-range">
                            {
                               monthRange.map((monthNumber, index) =>
                                <a key={index} 
                                    className={monthNumber === month ? "dropdown-item active" : "dropdown-item"}
                                    href="#"
                                    onClick={(e) => {this.selectMonth(e, monthNumber)}}
                                >
                                   {monthNumber}月
                                </a>
                                )
                            }
                        </div>
                    </div>
                }
            </div>
        );
    }
}

MonthPicker.propTypes = {
    year: PropTypes.number.isRequired,
    month: PropTypes.number.isRequired,
    onChange: PropTypes.func.isRequired
}
 
export default MonthPicker;