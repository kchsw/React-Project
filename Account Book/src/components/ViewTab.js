import React from 'react'
import Ionicon from 'react-ionicons'
import PropTypes from 'prop-types'
import { LIST_VIEW, CHART_VIEW } from '../utility'

const generateLinkClass = (current, view) => {
    return current === view? "nav-link active" : "nav-link"
}



const ViewTab = ({ activeTab, onTabChange }) => {
    const tabClick = (event, view) => {
        event.preventDefault();
        onTabChange(view)
    }
    return (
        <ul className="nav nav-tabs nav-fill my-4">
            <li className="nav-item list-item">
                <a className={generateLinkClass(activeTab, LIST_VIEW)} 
                    href="#"
                    onClick={(e) => {e.preventDefault(); onTabChange(LIST_VIEW)}}
                >
                    <Ionicon
                        className="rounded-circle mr-2"
                        fontSize="25px"
                        color={'#007bff'}
                        icon="ios-paper"
                    />
                    列表模式
                </a>
            </li>
            <li className="nav-item chart-item">
                <a className={generateLinkClass(activeTab, CHART_VIEW)} 
                    href="#"
                    onClick={(e) => {e.preventDefault(); onTabChange(CHART_VIEW)}}
                    // onClick={(e) => {tabClick(e, CHART_VIEW)}}
                >
                    <Ionicon
                        className="rounded-circle mr-2"
                        fontSize="25px"
                        color={'#007bff'}
                        icon="ios-pie"
                    />
                   图表模式
                </a>
            </li>
        </ul>
    )
}

ViewTab.propTypes = {
    activeTab: PropTypes.string.isRequired,
    onTabChange: PropTypes.func.isRequired,
}

export default ViewTab