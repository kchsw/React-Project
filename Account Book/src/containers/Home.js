import React, { Component } from 'react';
import logo from '../logo.svg';
import PriceList from '../components/PriceList'
import ViewTab from '../components/ViewTab'
import TotalPrice from '../components/TotalPrice'
import MonthPicker from '../components/MonthPicker'
import CreateBtn from '../components/CreateBtn'
import { LIST_VIEW, CHART_VIEW, TYPE_INCOME, TYPE_OUTCOME, parseDate, padLeft } from '../utility'

export  const categorys = {
    "1": {
            "id": 1,
            "name": "旅行",
            "type": "outcome",
            "iconName": "ios-plane"
    },
    "2": {
        "id": 2,
        "name": "理财",
        "type": "income",
        "iconName": "logo-yen"
    }

}
export const mockData = [
    {
      "id": 1,
      "title": "去云南旅行",
      "date": "2019/11/10",
      "price": 400,
      "cid": 1
    },
    {
      "id": 2,
      "title": "理财基金",
      "date": "2019/10/10",
      "price": 400,
      "cid": 2
    }
  ]

const newItem = {
        "id": 3,
        "title": "新添加项目",
        "date": "2019/11/10",
        "price": 600,
        "cid": 2
    }

    class Home extends Component {
        constructor(props){
            super(props)
            this.state = {  
                accountData: mockData,
                currentData: parseDate(),
                tabView: LIST_VIEW
            }
        }
        changeView = (view) => {
            this.setState({
                tabView: view
            })
        }
        changeDate = (year, month) => {
            this.setState({
                currentData: { year, month }
            })
        }
        modifyItem = (modifyiedItem) => {
            this.state.accountData.find(item => item.id === modifyiedItem.id).title = "nmsl"
            this.setState({
                accountData: this.state.accountData
            })
        }
        createItem = () => {
            this.setState({
                // accountData: mockData.concat(newItem)
                accountData: [newItem, ...this.state.accountData]
            })
        }
        deleteItem = (deletedItem) => {
            const filteredItems = this.state.accountData.filter(item => item.id !== deletedItem.id)
            this.setState({
                accountData: filteredItems
            })
        }
        render() { 
            let { accountData, currentData, tabView } = this.state
            accountData = accountData.map(item => {
                item.category = categorys[item.cid]
                return item
            }).filter(item => {
                return item.date.includes(`${currentData.year}/${padLeft(currentData.month)}`)
            })
            //这里直接对对象进行了操作
            // accountData.map(item => {
            //     item.category = categorys[item.cid]              
            // }) 
            let totalIncome = 0, totalOutcome = 0
            accountData.forEach(item => {
                item.category.type === TYPE_INCOME?
                totalIncome += item.price : totalOutcome += item.price
            })
            
            return (  
                <React.Fragment>
                    <header className="App-header">
                        <img src={logo} className="App-logo" alt="logo" />
                    </header> 
                    <div className="card" style={{width: '65%', margin: '0 auto'}}>
                        <div className="card-body">
                            <div className="row header">
                                <div className="col">
                                    <MonthPicker 
                                        year={currentData.year}
                                        month={currentData.month}
                                        onChange={this.changeDate}
                                    />
                                </div>
                                <div className="col d-flex justify-content-between align-items-center">
                                    <TotalPrice
                                        income={totalIncome}
                                        outcome={totalOutcome}
                                    />
                                </div>
                            </div>
                            <div className="content-area">
                                <ViewTab
                                    activeTab={tabView}
                                    onTabChange={this.changeView}
                                />
                                <CreateBtn onClick={this.createItem} />
                                {   tabView === LIST_VIEW &&
                                    <PriceList 
                                        items={accountData}
                                        onModifuItem={this.modifyItem}
                                        onDeleteItem={this.deleteItem}
                                    />
                                }
                                {
                                    tabView === CHART_VIEW &&
                                    <div>图表区域</div>
                                }
                            </div>
                        </div>
                    </div>
                </React.Fragment>
            );
        }
    }
   
  export default Home;