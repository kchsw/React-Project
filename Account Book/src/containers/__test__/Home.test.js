import React from 'react';
import { mount } from 'enzyme';
import Home from '../Home'

import PriceList from '../../components/PriceList'
import ViewTab from '../../components/ViewTab'
import MonthPicker from '../../components/MonthPicker'
import CreateBtn from '../../components/CreateBtn'
import { LIST_VIEW, CHART_VIEW, TYPE_INCOME, TYPE_OUTCOME, parseDate, padLeft } from '../../utility'

let wrapper

describe('test Home container component', () => {
    beforeEach(() => {
        wrapper = mount(<Home />)
    })
    it('snapshot', () => {
        expect(wrapper).toMatchSnapshot()
    })
    it('should render the default layout', () => {
        const currentDate = parseDate()
        expect(wrapper.find(PriceList).length).toEqual(1)
        expect(wrapper.find(ViewTab).props().activeTab).toEqual(LIST_VIEW)
        expect(wrapper.find(MonthPicker).props().year).toEqual(currentDate.year)
        expect(wrapper.find(PriceList).props().items.length).toEqual(1)
    })
    it('click the another view tab, should change the view', () => {
        wrapper.find('.chart-item a').first().simulate('click')
        expect(wrapper.find('.chart-item').length).toEqual(1)
        expect(wrapper.find(ViewTab).props().activeTab).toEqual(CHART_VIEW)
    })
    it('select another month should switch correct items', () => {
        wrapper.find('.dropdown-toggle').simulate('click')
        wrapper.find('.months-range .dropdown-item').at(9).simulate('click')
        expect(wrapper.find(MonthPicker).props().month).toEqual(10)
        expect(wrapper.find(PriceList).props().items.length).toEqual(1)
    })
    it('click the create button, should create the new item', () => {
        wrapper.find(CreateBtn).simulate('click')
        expect(wrapper.find(PriceList).props().items.length).toEqual(2)
    })
    it('modify item and delete item', () => {
        wrapper.find('.list-group-item a').first().simulate('click')
        expect(wrapper.find('.list-group-item .title-col').text()).toEqual('nmsl')
        expect(wrapper.find(PriceList).props().items[0].title).toEqual('nmsl')
        wrapper.find('.list-group-item a').last().simulate('click')
        expect(wrapper.find('.list-group-item').length).toEqual(0)
    })
})
