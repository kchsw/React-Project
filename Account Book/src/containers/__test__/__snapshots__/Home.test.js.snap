// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Home container component snapshot 1`] = `ReactWrapper {}`;
